### ISSUE TITLE
[Feature request] Backpropagation through Dataset API

### ISSUE THREAD

**Issue by Hoeze** [tensorflow/tensorflow#19556]
It would be cool if the Dataset API would support backpropagation.

Currenty I'm working on a statistical model with per-sample weights. 
Therefore, I have to pass sample indices through the Dataset API and index the weights manually.

[See my question (especially Allen Lavoie's comment) on stackoverflow.](https://stackoverflow.com/questions/50155021/tensorflow-dataset-api-gradient-is-none?noredirect=1#comment87464556_50155021)

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.8.0
- **Python version**: 3.6.5
- **Bazel version (if compiling from source)**: none
- **GCC/Compiler version (if compiling from source)**: none
- **CUDA/cuDNN version**: none
- **GPU model and memory**: none
- **Exact command to reproduce**: none

**Comment by aselle** [comment]
Derek @mrry, please take a look and see if this feature request makes sense to your dev plan.

**Comment by mrry** [comment]
This sounds like it could be useful, but it's not something we're planning to work on in the near term. The major complexity arises from our main gradients implementation (`tf.gradients()`), which only works on purely function ops, and is currently unaware of stateful operations like `Iterator.get_next()`. Perhaps the `tfe.GradientTape`–based gradients used in eager execution mode (and which work in graph mode) could more easily be extended to support this, but it's not something anybody's looked at yet.

I'll throw this open to community contributions, since it would be great if someone took the time to figure out how this could work.
